### PR TITLE
AGENT-1454: Enforce APIServer TLS profile on IRI registry

### PR DESCRIPTION
**Pull request by rwsu**
**- What I did**

Read the cluster's APIServer TLS security profile and apply it to the IRI registry's REGISTRY_HTTP_TLS_MINIMUMTLS and REGISTRY_HTTP_TLS_CIPHERSUITES environment variables. When no profile is set, the Intermediate profile (TLS 1.2) is used as the default.
    
The implementation:
- Adds an APIServer informer to the IRI controller so TLS profile changes trigger reconciliation
- Converts OpenShift TLS version strings (e.g. "VersionTLS12") to the registry format (e.g. "tls1.2") programmatically, supporting future TLS versions without code changes
- Filters cipher suites to only pass configurable ones (TLS 1.2 and below) to the registry, since TLS 1.3+ ciphers are fixed by the protocol. Uses Go's CipherSuite.SupportedVersions for future-proof detection
- Updates bootstrap path to pass TLS profile through
- Adds unit test for Modern TLS profile (TLS 1.3)
- Adds e2e test that dynamically verifies TLS enforcement based on the cluster's active profile
- Add a NodeDisruptionPolicy to allow rebootless updates to IRI TLS setting. Nodes are still rebooted when the cluster's TLSSecurityProfile is changed because of changes to the kubelet config and kube-rbac-proxy-crio static pod do not have a NodeDisruptionPolicy that prevent reboots.
- 
**- How to verify it**

Update the cluster TLSSecurtyProfile to Modern (TLS 1.3)

`oc patch apiserver cluster --type merge -p '{"spec":{"tlsSecurityProfile":{"type":"Modern","modern":{}}}}'    `

Check /etc/systemd/system/iri-registry.service is updated to set TLS1.3 as mininum TLS version supported.

`-e REGISTRY_HTTP_TLS_MINIMUMTLS=tls1.3`

**- Description for the changelog**

Enforce APIServer TLS security profile on the IRI registry.

## Release Notes

* **New Features**
  * Internal container registry now respects cluster TLS security profiles, automatically configuring minimum allowed TLS version and cipher suites based on system-wide policy settings.

* **Bug Fixes**
  * Enhanced service disruption handling policies to include registry service restart and daemon reload operations.

* **Tests**
  * Added end-to-end test suite to verify TLS profile enforcement on registry connections.

### CODE REVIEW

**Comment by rwsu** [comment]
/verified by @rwsu tested IRI registry TLS settings are updated when cluster level APIServer 's TLSSecurityProfile is changed. Also have e2e test to verify enforcement of currently active profile. And unit test that verifies profile changes produces the correct systemd unit content.

File: pkg/apihelpers/apihelpers.go [hunk 1]
```diff
@@ -108,28 +108,18 @@ var (
...
-							ServiceName: "kubelet.service",
-						},
-					},
-				},
-			},
+		},
+		Units: []opv1.NodeDisruptionPolicyStatusUnit{
 			{
-				Path: constants.KubeletCrioImageCredProviderConfPath,
+				Name: opv1.NodeDisruptionPolicyServiceName(constants.IRIRegistryServiceName),
```
Comment (andfasano) [inline pkg/apihelpers/apihelpers.go:173]: note: also @bfournie had a PR hitting here https://github.com/openshift/machine-config-operator/pull/5683/changes#r2983557894

File: pkg/controller/internalreleaseimage/internalreleaseimage_bootstrap_test.go [hunk 2]
```diff
@@ -8,7 +8,7 @@ import (
 )
 
 func TestRunInternalReleaseImageBootstrap(t *testing.T) {
-	configs, err := RunInternalReleaseImageBootstrap(&mcfgv1alpha1.InternalReleaseImage{}, iriCertSecret().obj, cconfig().obj)
+	configs, err := RunInternalReleaseImageBootstrap(&mcfgv1alpha1.InternalReleaseImage{}, iriCertSecret().obj, cconfig().obj, nil)
```
Comment (andfasano) [inline pkg/controller/internalreleaseimage/internalreleaseimage_bootstrap_test.go]: What about also adding a test to cover the case where the apiserver is not nil?

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller.go [hunk 3]
```diff
@@ -341,8 +372,29 @@ func (ctrl *Controller) syncInternalReleaseImage(key string) error {
 		return fmt.Errorf("could not get Secret %s: %w", ctrlcommon.InternalReleaseImageTLSSecretName, err)
 	}
 
+	// Fetch the APIServer TLS profile. If not found, use nil (defaults to Intermediate).
```
Comment (andfasano) [inline pkg/controller/internalreleaseimage/internalreleaseimage_controller.go]: For the sake of maintenance/readability  I think it's important to keep the root level of `syncInternalReleaseImage` as much as possible short and clean. Could you please refactor this block within a `getTLSProfile()` method?

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller_test.go
Comment (andfasano) [inline pkg/controller/internalreleaseimage/internalreleaseimage_controller_test.go:1]: Nice!

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller.go [hunk 4]
```diff
@@ -341,8 +372,29 @@ func (ctrl *Controller) syncInternalReleaseImage(key string) error {
...
+	apiServer, err := ctrl.apiServerLister.Get(ctrlcommon.APIServerInstanceName)
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("could not get APIServer: %w", err)
+	}
+	var tlsProfile *configv1.TLSSecurityProfile
+	if apiServer != nil {
+		tlsProfile = apiServer.Spec.TLSSecurityProfile
```
Comment (andfasano) [inline pkg/controller/internalreleaseimage/internalreleaseimage_controller.go]: q: will be this field effectively available in a GA cluster? IIUC now it's still in TP

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller.go [hunk 5]
```diff
@@ -341,8 +372,29 @@ func (ctrl *Controller) syncInternalReleaseImage(key string) error {
...
+	}
+
+	profileType := "nil (defaulting to Intermediate)"
+	if tlsProfile != nil {
+		profileType = string(tlsProfile.Type)
+	}
+	tlsMinVersion, tlsCipherSuites := registryTLSFromProfile(tlsProfile)
```
Comment (andfasano) [inline pkg/controller/internalreleaseimage/internalreleaseimage_controller.go]: Is this really required here? IINW `registryTLSFromProfile` could be moved as part of the Renderer type and the trace log moved also there

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller.go (hunk 4 above)
Comment (rwsu) [reply pkg/controller/internalreleaseimage/internalreleaseimage_controller.go]: APIServer.Spec.TLSSecurityProfile is part of config.openshift.io/v1 and has been in the GA API since OCP 4.2. Full support for the Modern profile (TLS 1.3) was added in OCP 4.19, but we handle all profiles (Old, Intermediate, Modern, Custom) and default to Intermediate when unset, so the field is usable across all supported GA clusters.

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller.go (hunk 3 above)
Comment (rwsu) [reply pkg/controller/internalreleaseimage/internalreleaseimage_controller.go]: Done.

File: pkg/controller/internalreleaseimage/internalreleaseimage_bootstrap_test.go (hunk 2 above)
Comment (rwsu) [reply pkg/controller/internalreleaseimage/internalreleaseimage_bootstrap_test.go:13]: Done — added TestRunInternalReleaseImageBootstrap_WithTLSProfile which passes a Modern TLS profile and verifies the rendered unit contains tls1.3 with no cipher suites.

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller.go (hunk 5 above)
Comment (rwsu) [reply pkg/controller/internalreleaseimage/internalreleaseimage_controller.go]: Done — removed the registryTLSFromProfile call and log from syncInternalReleaseImage. The log is now in Renderer.newRenderContext() at V(4), which is the natural home for it since that's where registryTLSFromProfile is already called.

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller.go [hunk 6]
```diff
@@ -513,6 +547,24 @@ func (ctrl *Controller) addFinalizerToInternalReleaseImage(iri *mcfgv1alpha1.Int
...
+// honor the cluster TLS profile regardless of APIServer.Spec.TLSAdherence.
+// The tlsAdherence field exists to provide a migration path for existing components
+// that historically ignored the cluster profile; that concern does not apply here.
+func (ctrl *Controller) getTLSProfile() (*configv1.TLSSecurityProfile, error) {
+	apiServer, err := ctrl.apiServerLister.Get(ctrlcommon.APIServerInstanceName)
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, fmt.Errorf("could not get APIServer: %w", err)
+	}
+	if apiServer != nil {
+		return apiServer.Spec.TLSSecurityProfile, nil
```
Comment (andfasano) [inline pkg/controller/internalreleaseimage/internalreleaseimage_controller.go:582]: @rwsu can you please remind me if that field is supported in 4.22? We don't plan to backport it, but just in case

**Comment by andfasano** [comment]
/lgtm
/hold 

Hold to be removed once we agreed to include it or not in the eventual 4.22 backport plan

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller.go (hunk 6 above)
Comment (rwsu) [reply pkg/controller/internalreleaseimage/internalreleaseimage_controller.go:582]: Yes the field is supported in 4.22. APIServer.Spec.TLSSecurityProfile is part of config.openshift.io/v1 and has been in the GA API since approximately OCP 4.2 (https://github.com/openshift/api/pull/452).

**Comment by andfasano** [comment]
/hold cancel

As previously discussed there shouldn't be no harm in backporting also this feature in 4.22

**Comment by rwsu** [comment]
/verified by TestIRIController_VerifyTLSProfileEnforced and  "apply Modern TLS profile from APIServer" unit test

File: test/e2e-iri/iri_test.go [hunk 7]
```diff
@@ -256,6 +256,110 @@ func TestIRIAuth_AuthenticatedRequestSucceeds(t *testing.T) {
 	require.Equal(t, "200", statusCode, "authenticated request should succeed")
 }
 
+// TestIRIController_VerifyTLSProfileEnforced verifies that the IRI registry
+// enforces the TLS minimum version derived from the cluster's APIServer TLS profile.
+// It connects with the expected minimum version (should succeed) and one version
+// below it (should be rejected). Uses ExecCmdOnNode to run curl on the node
+// directly, since port 22625 is not accessible from the CI test runner.
+func TestIRIController_VerifyTLSProfileEnforced(t *testing.T) {
```
Comment (pablintino) [inline test/e2e-iri/iri_test.go:264]: Cool, I was about to say that this test is not changing the TLS config of the cluster and thus both cluster and your internal default are the same leading to lack of coverage, but I saw that you have a few UTs that cover that.

File: pkg/controller/internalreleaseimage/internalreleaseimage_renderer.go [hunk 8]
```diff
@@ -118,19 +126,102 @@ func (r *Renderer) newRenderContext() (*renderContext, error) {
...
 		IriHtpasswd:         iriHtpasswd,
+		TLSMinVersion:       tlsMinVersion,
+		TLSCipherSuites:     tlsCipherSuites,
 	}, nil
 }
 
+// registryTLSFromProfile converts an OpenShift TLSSecurityProfile to the
+// Distribution registry's environment variable values for REGISTRY_HTTP_TLS_MINIMUMTLS
+// and REGISTRY_HTTP_TLS_CIPHERSUITES.
+func registryTLSFromProfile(profile *configv1.TLSSecurityProfile) (minVersion, cipherSuites string) {
```
Comment (pablintino) [inline pkg/controller/internalreleaseimage/internalreleaseimage_renderer.go:156]: Returning a "yaml" formated string for `cipherSuites` is a bit ugly. Why not returning the raw slice and formating it with the Got emplate?

File: pkg/controller/internalreleaseimage/internalreleaseimage_controller.go [hunk 9]
```diff
@@ -532,6 +566,24 @@ func (ctrl *Controller) addFinalizerToInternalReleaseImage(iri *mcfgv1alpha1.Int
 	return err
 }
 
+// getTLSProfile fetches the cluster APIServer TLS security profile.
+// Returns nil (which defaults to Intermediate) if the APIServer object is not found.
```
Comment (pablintino) [inline pkg/controller/internalreleaseimage/internalreleaseimage_controller.go:570]: nit: `(which defaults to Intermediate)` that's a decission the caller should make, so I don't think stating `which defaults to Intermediate` here does add value.

File: pkg/controller/internalreleaseimage/internalreleaseimage_renderer_test.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+		},
+		{
+			name:     "single digit defaults to tls1.2",
+			input:    "VersionTLS1",
+			expected: "tls1.2",
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expected, openShiftTLSVersionToRegistryVersion(tc.input))
```
Comment (pablintino) [inline pkg/controller/internalreleaseimage/internalreleaseimage_renderer_test.go:63]: nit: Testing package private functions is not something usually recommended and usually considering whitebox testing.
